File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/QuitRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.constant.ResultType;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.QuitRequest;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.dto.websocket.outbound.QuitResponse;
import net.prostars.messagesystem.service.ChannelService;
import net.prostars.messagesystem.service.ClientNotificationService;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class QuitRequestHandler implements BaseRequestHandler<QuitRequest> {

  private final ChannelService channelService;
  private final ClientNotificationService clientNotificationService;

  public QuitRequestHandler(
      ChannelService channelService, ClientNotificationService clientNotificationService) {
    this.channelService = channelService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, QuitRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());

    ResultType result;
    try {
      result = channelService.quit(request.getChannelId(), senderUserId);
    } catch (Exception ex) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.QUIT_REQUEST, ResultType.FAILED.getMessage()));
      return;
    }

    if (result == ResultType.SUCCESS) {
      clientNotificationService.sendMessage(
          senderSession, senderUserId, new QuitResponse(request.getChannelId()));
    } else {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.QUIT_REQUEST, result.getMessage()));
    }
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/QuitRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.ResultType
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.QuitRequest
import com.yscorp.withpush.messagesystem.service.ChannelService
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import net.prostars.messagesystem.constant.IdKey
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class QuitRequestHandler(
    private val channelService: ChannelService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<QuitRequest> {

    /**
     * 사용자의 채널 퇴장 요청을 처리합니다.
     * 성공 시 QuitResponse, 실패 시 ErrorResponse 전송.
     */
    override fun handleRequest(senderSession: WebSocketSession, request: QuitRequest) {
        val senderUserId = senderSession.attributes[IdKey.USER_ID.value] as? UserId ?: return

        val result = runCatching {
            channelService.quit(request.channelId, senderUserId)
        }.getOrElse {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.QUIT_REQUEST, ResultType.FAILED.message)
            )
            return
        }

        if (result == ResultType.SUCCESS) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                QuitResponse(request.channelId)
            )
        } else {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.QUIT_REQUEST, result.message)
            )
        }
    }
}
